### PR TITLE
Add `stopOnDomReady` option, remove `.cancel()`, add `.stop()`

### LINKED ISSUE #8: Automatically stop on dom ready
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/element-ready/issues/8)

> ```js
>	const waiting = elementReady(selector);
>
>	// Don't check ad-infinitum
>	domLoaded.then(() => requestAnimationFrame(() => waiting.cancel()));
>
>	// If cancelled, return null like a regular querySelector() would
>	return waiting.catch(() => null);
> ```
> Note to self: This should really be an option in element-ready.
> @sindresorhus on https://github.com/sindresorhus/refined-github/pull/605#discussion_r126411888

> There is a $40.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/element-ready/issues/8).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/element-ready/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by bfred-it**
Requires semver major because it changes the default behavior.

- [x] Approval
- [x] `index.test-d.ts` with `null`, why is it failing?
- [x] Docs

Fixes #8

### CODE REVIEW

File: index.test-d.ts [hunk 1]
```diff
@@ -6,6 +6,7 @@ const promise = elementReady('#unicorn');
 elementReady('#unicorn', {target: document});
 elementReady('#unicorn', {target: document.documentElement});
 
+expectType<PCancelable<null>>(promise);
 expectType<PCancelable<Element>>(promise);
```
Comment (sindresorhus) [inline index.test-d.ts]:
```suggestion
expectType<PCancelable<Element | null>>(promise);
```

Instead of line 9.

File: test.js [hunk 2]
```diff
@@ -68,11 +73,38 @@ test('check if different elements ready inside different targets with same selec
 	t.is(element2.id, 'unicorn2');
 });
 
+test('check if element ready after dom loaded', async t => {
+	const elementCheck = elementReady('#bio', {
+		cancelOnDomLoaded: true
+	});
+
+	delay(50000).then(() => {
```
Comment (sindresorhus) [inline test.js]: Use await. Like: https://github.com/sindresorhus/element-ready/commit/513139a880cc2808666ebfae6c2d47b89b624379#diff-1dd241c4cd3fd1dd89c570cee98b79ddR16

**Comment by sindresorhus** [comment]
> Requires semver major because it changes the default behavior.

Good timing. I already need to do a major release for other things in master.

> Approval

👍 

> index.test-d.ts with null, why is it failing?

I commented inline.

---

Can you fix the merge conflict?

---

I think we should use `undefined` instead of `null`.

**Comment by bfred-it** [comment]
Null is what querySelector returns, so I thought of keeping it consistent

**Comment by sindresorhus** [comment]
I’m trying to get rid of `null` in my packages. We can’t change the browser APIs, but at least we can make our code better.

**Comment by bfred-it** [comment]
What’s wrong with null?

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/meta/issues/7

File: test.js (hunk 2 above)
Comment (bfred-it) [reply test.js]: Can't, it needs to happen after line 87 (`await elementCheck`), and we don't actually need to await it, `elementCheck` is supposed to "cancel on dom ready" and return `null` _before_ that code is run, hence the long timer.

File: index.test-d.ts (hunk 1 above)
Comment (bfred-it) [reply index.test-d.ts]: Done with `| undefined`

**Comment by bfred-it** [comment]
Are only docs  missing here?

**Comment by sindresorhus** [comment]
Yes

**Comment by bfred-it** [comment]
Good to go, however we should probably rename it:

- the promise isn't actually `cancel`led, but resolved to `null`
- the "dom ready" event is not commonly known as "dom loaded"

So perhaps it should be `stopOnDomReady` or `untilDomReady`

**Comment by sindresorhus** [comment]
I'm starting to think that `.cancel()` should be changed too. Since there are many paths that could end up with `null` now, consumers needs to handle it anyway, so `.cancel()` could just resolve to `null` instead of rejecting. So maybe we should rename `.cancel()` to `.stop()` too?

I like `stopOnDomReady`.

File: index.js [hunk 3]
```diff
@@ -20,25 +23,32 @@ const elementReady = (selector, options) => {
...
 			cancelAnimationFrame(rafId);
-			cache.delete([target, selector], promise);
+			cache.delete(cacheKeys, promise);
 		});
 
+		if (cancelOnDomLoaded) {
+			domLoaded.then(() => {
+				cancelAnimationFrame(rafId);
+				resolve();
+			});
```
Comment (sindresorhus) [inline index.js]: I know you hate this, but I prefer:

```js
(async () => {
	await domLoaded;
	cancelAnimationFrame(rafId);
	resolve();
})();
```

**Comment by bfred-it** [comment]
> So maybe we should rename `.cancel()` to `.stop()` too?

So I should drop `p-cancelable` since it wouldn't be a regular cancelable promise.

File: index.js (hunk 3 above)
Comment (bfred-it) [reply index.js]: What's the problem with `.then`?

File: test.js [hunk 4]
```diff
@@ -71,40 +75,69 @@ test('check if different elements ready inside different targets with same selec
...
+test('ensure different promises are returned on second call with the same selector when first was stopped', async t => {
+	const elementCheck1 = elementReady('.unicorn', {stopOnDomReady: false});
+
+	elementCheck1.stop();
 
-	elementCheck1.cancel();
+	const elementCheck2 = elementReady('.unicorn', {stopOnDomReady: false});
 
-	const elementCheck2 = elementReady('.unicorn');
+	t.is(await elementCheck1, undefined);
```
Comment (bfred-it) [inline test.js]: Does this do what you wanted? Or does it no longer apply?

**Comment by sindresorhus** [comment]
> So I should drop p-cancelable since it wouldn't be a regular cancelable promise.

Yes

File: index.d.ts [hunk 5]
```diff
@@ -10,6 +9,10 @@ declare namespace elementReady {
 		*/
 		readonly target?: Element | Document;
 	}
+
+	type PStoppable<T> = Promise<T> & {
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	type StoppablePromise<T> = Promise<T> & {
```

File: readme.md [hunk 6]
```diff
@@ -47,16 +47,21 @@ Default: `document`
...
+Default: `true`
+
+Automatically stops checking for the element to be ready after the DOM ready event.
+
+### elementReadyPromise#stop()
 
 Type: `Function`
 
-Stops checking for the element to be ready. The cancelation is synchronous.
+Stops checking for the element to be ready. The stop is synchronous.
```
Comment (sindresorhus) [inline readme.md]: Should document that it then resolves with `undefined`.

File: test.js (hunk 4 above)
Comment (sindresorhus) [reply test.js]: If you mean resolving to `undefined` when `.stop()` is called, then yes, it's what I wanted.

File: index.js [hunk 7]
```diff
@@ -1,47 +1,50 @@
...
 const ManyKeysMap = require('many-keys-map');
+const domLoaded = require('dom-loaded');
+const pDefer = require('p-defer');
 
 const cache = new ManyKeysMap();
+const elementReady = (selector, options = {}) => {
+	const {
+		target = document,
+		stopOnDomReady = true
+	} = options;
```
Comment (bfred-it) [inline index.js]: Thoughts on this pattern instead?


```js
const elementReady = (selector, {
	target = document,
	stopOnDomReady = true
} = {}) => {
	// Code
};
```

Comment (sindresorhus) [reply index.js]: Sure, that's fine too. It just has a tendency to become kinda messy when you have a lot of options.

**Pull request merged**